src/schemes/fluid/implicit_incompressible_sph/system.jl

```diff
@@ -392,51 +389,102 @@ function calculate_diagonal_elements!(a_ii, system, neighbor_system::AbstractBou
 
         # Contribution to the diagonal elements without d_ji value (see eq. 16 in Ihmsen et al. (2013))
         a_ii[particle] += m_b * dot(d_ii_, grad_kernel)
+
+        # Calculate the predicted velocity differences
+        advection_velocity_diff = predicted_velocity(system, particle) -
+                                  predicted_velocity(neighbor_system, neighbor)
+
+        # Compute \rho_adv in eq. 4 in Ihmsen et al. (2013)
+        predicted_density[particle] += time_step * m_b *
+                                       dot(advection_velocity_diff, grad_kernel)
     end
 end
 
 # Calculate pressure values with iterative pressure solver (relaxed Jacobi scheme)
-function pressure_solve!(system, v, u, v_ode, u_ode, semi, t)
-    (; reference_density, max_error, min_iterations, max_iterations, time_step) = system
+function pressure_solve!(semi, v_ode, u_ode, t)
+    foreach_system(semi) do system
+        initialize_pressure!(system, semi)
+    end
+
+    # Determine global iteration and error constraints across all IISPH systems
+    min_iters = maximum(minimum_iisph_iterations, semi.systems)
+    max_iters = minimum(maximum_iisph_iterations, semi.systems)
+    max_err = minimum(maximum_iisph_error, semi.systems)
 
-    initialize_pressure!(system, semi)
-    l = 1
-    terminate = false
     # Convert relative error in percent to absolute error
-    eta = max_error * reference_density / 100
+    eta = max_err / 100
```

Comment:
Now that this is not identical to the eta from the paper, I would prefer a different name. Maybe `max_err` and rename the one above to `max_error_percent`.
```suggestion
    max_err = max_err_percent / 100
```